src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1404,163 +1390,341 @@ object Desugar {
+            *         k := key
+            *         v := x[k] // [v]
+            *         exhale acc(x, 1/100000)
+            *         <body>
+            *         inhale acc(x, 1/100000)
+            *         visited := visited union Set(k)
             * 
-            * if (0 <= expIndex && expIndex < len(c)) { // [v]
-            *     expValue = c[expIndex]
-            * }
-            *
-            * <invariants were expIndex is replaced with f(i, len(c) > 0) and j with f(j, len(c) > 0)>
-            * invariant len(c) > 0 ==> 0 <= i0 && i0 <= len(c)               // these invariants might require access to something
-            * invariant len(c) > 0 ==> expIndex == i0                        // as expIndex can be an indexing operation for example
-            * invariant len(c) > 0 ==> 0 <= expIndex && expIndex <= len(c)   // so they are added after the user invariants
-            * invariant len(c) > 0 ==> (0 <= expIndex && expIndex < len(c) ==> expValue == c[expIndex]) // [v]
-            * for expIndex < len(c) {
-            *     <body>
-            *     i0 += 1
-            *     expIndex = i0
-            *     if (0 <= expIndex && expIndex < len(c)) { // [v]
-            *         expValue = c[i]
-            *     }
-            * }
-            *
-            * In the case where the value expression 'expValue' is missing all the code annotated with [v]
+            * In the case where the value variable 'v' is missing all the code annotated with [v]
             * is omitted
             */
-          case n@PAssForRange(range, ass, spec, body) =>
+          case n@PShortForRange(range, shorts, _, spec, body) if underlyingType(info.typ(range.exp)).isInstanceOf[MapT] =>
+            // is a block as it introduces new variables
             unit(block(for {
               exp <- goE(range.exp)
-              typ = underlyingType(exp.typ) match {
-                case slice : in.SliceT => slice
-                case array : in.ArrayT => array
-                case _ : in.MapT => violation("Maps are not supported yet in range")
-                case _ : in.StringT => violation("Strings are not supported yet in range")
-                case t => violation(s"Range not applicable to type $t")
-              }
-              indexLeft <- goL(ass(0))
-              rangeSrc = meta(range, info)
-              indexSrc = meta(ass(0), info)
-              rangeExpSrc = meta(range.exp, info)
 
-              copiedVar <- freshDeclaredExclusiveVar(exp.typ, n, info)(rangeExpSrc)
-              copyAss = singleAss(in.Assignee.Var(copiedVar), exp)(rangeExpSrc)
-              length = in.Length(copiedVar)(rangeSrc)
+              keyType = exp.typ.asInstanceOf[in.MapT].keys.withAddressability(Addressability.exclusiveVariable)
+              valType = exp.typ.asInstanceOf[in.MapT].values.withAddressability(Addressability.exclusiveVariable)
+              visType = in.SetT(keyType, Addressability.exclusiveVariable)
```

Comment:
I think that the declaration of `keyType` and `valType` here could be a bit safer by using a case analysis:
```
              val (keyType, valType) = underlyingType(exp.typ) match {
                case in.MapT(k, v, _) => (k.withAddressability(???), v.withAddressability(???))
                case _ => violation(???)
              }
```
Also, notice that I'm using `underlyingType` here applied to `exp.type`. I'm not sure if we need `underlyingType` here but my impression is that, without it, we may run into exceptions when ranging over a defined type.  I also wonder why we are creating `exclusive` types no matter what. Shouldn't it depend on the addressability vector in `n`?

Reply:
I think that there are other places where my comments may apply

Reply:
At that point though that I am doing the cast, we have already pattern matched on the type of the expression and we know it is a map.